src/window/update_loop.rs

```diff
@@ -269,61 +318,89 @@ impl UpdateLoop {
-                    self.process_buffered_draw_commands(window_wrapper);
-                } else {
-                    tracy_zone!("redraw requested");
-                    // The OS itself asks us to redraw, so we need to prepare first
-                    self.should_render = ShouldRender::Immediately;
-                }
+                #[cfg(target_os = "macos")]
+                self.menu.ensure_menu_added();
             }
             _ => {}
         }
 
-        if self.pending_render && matches!(&event, Event::UserEvent(UserEvent::DrawCommandBatch(_)))
-        {
-            // Buffer the draw commands if we have a pending render, we have already decided what to
-            // draw, so it's not a good idea to process them now.
-            // They will be processed immediately after the rendering.
-            self.pending_draw_commands.push(event);
-        } else if window_wrapper.handle_event(event) {
-            // But we need to handle other events (in the if statement itself)
-            // Also schedule a render as soon as possible
+        if self.window_wrapper.handle_window_event(event) {
             self.should_render = ShouldRender::Immediately;
         }
+        self.schedule_next_event(event_loop);
+    }
 
-        #[cfg(feature = "profiling")]
-        self.should_render.plot_tracy();
+    fn user_event(&mut self, event_loop: &ActiveEventLoop, event: UserEvent) {
+        tracy_zone!("user_event");
+        match event {
+            UserEvent::NeovimExited => {
+                save_window_size(&self.window_wrapper);
+                event_loop.exit();
+            }
+            UserEvent::RedrawRequested => {
+                self.redraw_requested();
+            }
+            UserEvent::DrawCommandBatch(batch) if self.pending_render => {
+                // Buffer the draw commands if we have a pending render, we have already decided what to
+                // draw, so it's not a good idea to process them now.
+                // They will be processed immediately after the rendering.
+                self.pending_draw_commands.push(batch);
+            }
+            _ => {
+                self.window_wrapper.handle_user_event(event);
+                self.should_render = ShouldRender::Immediately;
+            }
+        }
+        self.schedule_next_event(event_loop);
+    }
+
+    fn about_to_wait(&mut self, event_loop: &ActiveEventLoop) {
+        tracy_zone!("about_to_wait");
+        self.prepare_and_animate();
+        self.schedule_next_event(event_loop);
+    }
+
+    fn resumed(&mut self, event_loop: &ActiveEventLoop) {
+        tracy_zone!("resumed");
+        self.create_window_allowed = true;
+        self.schedule_next_event(event_loop);
+    }
 
-        ControlFlow::WaitUntil(self.get_event_deadline())
+    fn exiting(&mut self, event_loop: &ActiveEventLoop) {
+        tracy_zone!("exiting");
+        self.window_wrapper.exit();
```

Comment:
I don't like this this, but the window and skia has to be destroyed before the event loop finishes to avoid this crash:
* https://github.com/neovide/neovide/issues/2223

It would be nice if the event loop took ovnership of the `UpdateLoop` and destroyed it automatically. But unfortunately `event_loop.run_app` takes a reference.